### ISSUE TITLE
Jacobian and grad after getindex

### ISSUE THREAD

**Issue by baggepinnen** [FluxML/Flux.jl#101]
I am missing calculation of jacobians
```julia
julia> using Flux

julia> using Flux: back!

julia> x = randn(2)
2-element Array{Float64,1}:
 -0.488323
  0.143252

julia> y = Dense(2,2)(x)

Tracked 2-element Array{Float64,1}:
 -0.0087318
  0.0192173

julia> back!(y)
ERROR: MethodError: no method matching back!(::TrackedArray{…,Array{Float64,1}})
Closest candidates are:
  back!(::Flux.Tracker.TrackedArray, ::Any) at /home/fredrikb/.julia/v0.6/Flux/src/tracker/back.jl:39
  back!(::Flux.Tracker.TrackedArray{T,0,A} where A where T) at /home/fredrikb/.julia/v0.6/Flux/src/tracker/back.jl:43
```

Furthermore, the following confuses me
```julia
julia> x = param(randn(2))
Tracked 2-element Array{Float64,1}:
  0.567788
 -0.32716 

julia> y = Dense(2,2)(x)
Tracked 2-element Array{Float64,1}:
  0.00648598
 -0.00119976

julia> back!(y[1])

julia> x.grad
2-element Array{Float64,1}:
 -0.00194926 
 -0.000263109
```

while the following returns an error
```julia
julia> x[1].grad
ERROR: UndefRefError: access to undefined reference
```
It seems the gradient information is lost when accessing an element

**Comment by MikeInnes** [comment]
For `back!`, if the object isn't a scalar, you need to provide a gradient. e.g. `back!(y, [1, 2])`. We need a much better error message here though.

To get a gradient of an array element the right way would be `x.grad[1]`. We can't carry gradient information forward when you operate on a TrackedArray.

**Comment by baggepinnen** [comment]
I'm trying to implement a simple `jacobian` function but really can't figure out how the gradient is calculated and accessed. The example below tries to model a time series `x[k+1] = Ax[k]`. I want to recover the `A` from the neural network by calculating the jacobian of the vector-valued output with respect to the vector-valued input. How do I calculate the gradient of a TrackedArray wrt an `Array/TrackedArray`? My attempt below does not work, but might highlight my misunderstanding.

```julia
    using Flux
    using Flux: back!, truncate!, treelike, train!, mse
    N      = 300
    n      = 2

    function generate_data()
        A      = [0.999 1; 0 0.9]
        x0     = randn(n)
        x      = zeros(n,N)
        x[:,1] = x0
        for i = 1:N-1
                x[:,i+1] = -A*x[:,i] + 0.1randn(n)
        end

        y = x[:,2:end]
        x = x[:,1:end-1]
        x,y,A
    end

x,y,A = generate_data()
np = 17
m = Chain(
    Dense(n,np,swish),
    Dense(np,np,swish),
    Dense(np,np,swish),
    Dense(np,n)
)
loss = (x,y) -> Flux.mse(m(x), y)

function fit_model(m, loss, x, y)
    dataset = Iterators.repeated((x, y), 2000)
    opt = ADAM(params(m), 0.01)
    train!(loss, dataset, opt)
    loss(x,y).data[1]
end

fit_model(m, loss, x, y)

function jacobian(x,m)
    xp = param(x)
    y  = m(xp) # Use model to predict
    @show xp.grad
    m  = length(y)
    n  = length(x)
    J  = zeros(n,m)
    for i = 1:m
        back!(y[i]) # What to write as second argument? Can the loop be avoided 
        J[:,i] = xp.grad
    end
    J
end
jacobian(x[:,100],m) # Should equal == A but doesn't
```

**Comment by baggepinnen** [comment]
Okay, the following produces a reasonable jacobian
```julia
function jacobian(x,m)
    xp = param(x)
    y  = m(xp) # Use model to predict, cant figure out the size of the model output in any other way?
    k  = length(y)
    n  = length(x)
    J  = zeros(n,k)
    for i = 1:k
        Flux.back!(y[i])
        J[:,i] = xp.grad
        xp.grad .*= 0 # This is improtant
    end
    J'
end
jacobian(x[:,200],m)
```

**Comment by MikeInnes** [comment]
Closed with #124.
